unstructured/file_utils/metadata.py

```diff
@@ -150,3 +153,43 @@ def _get_exif_datetime(exif_dict: Dict[str, Any], key: str) -> Optional[datetime
     # using the standard EXIF datetime format
     except ValueError:
         return None
+
+
+def apply_lang_metadata(
+    elements: Iterable[Element],
+    languages: List[str],
+    detect_language_per_element: bool = False,
+) -> Iterator[Element]:
+    """Detect and apply metadata.languages to each element in `elements`."""
+    # -- Note this function has a stream interface, but reads the full `elements` stream into memory
+    # -- before emitting the first updated element as output.
+
+    # Skip language detection for partitioners that use other partitioners.
+    # For example, partition_msg relies on partition_html and partition_text, but the metadata
+    # gets overwritten after elements have been returned by _html and _text,
+    # so `languages` would be detected twice.
+    if languages == [""]:
+        yield from elements
+        return
+
+    if not isinstance(elements, List):
+        elements = list(elements)
+
+    full_text = " ".join(e.text for e in elements if hasattr(e, "text"))
```

Comment:
So the `full_text` is being concatenated in post-processing for every doctype - is this better (or at least not too inefficient comparatively) than using the full text that exists when initially retrieved from the document (ex. partition/text.py)? I understand this needs to be done for certain filetypes that never have the full text available (like docx), but am wondering about the cases where we do have full text available in the partitioner.

Reply:
A couple of reasons seemed to justify this:
- it seemed reasonable to have all the partitioners use the same approach for detecting languages rather than make the code more complicated by having some use one approach and others use another approach--I did try to keep the pattern you initially used at first, but it just ended up being a lot easier to make them uniform. I believe @scanny also told me at some point that the additional iteration over elements to get the document text wasn't going to add a significant computational load (informally running tests didn't seem to take longer at the least) and it was better to prioritize composability over efficiency in this case.
- Detecting the languages after the elements had been created also made it a lot easier to detect language at the element level if multiple languages were detected at the document level or the `detect_language_per_element` param is set to `True` (and again not have to have very different approaches for the different partitioners)

Reply:
That makes sense, since tests didn't show too much extra load, the consistency across partitioners is worth it. Thanks!